metronome/checkpointing/service/test/src/io/iohk/metronome/checkpointing/service/CheckpointingServiceProps.scala

```diff
@@ -0,0 +1,407 @@
+      ) =
+        Random
+          .shuffle(validationBatch)
+          .map(b =>
+            Task.defer {
+              for {
+                _ <- validating.set(true)
+                e <- executing.get
+                _ <- achievedPar.update(_ || e)
+                r <- checkpointingService.validateBlock(b)
+                _ <- validating.set(false)
+              } yield (r.getOrElse(false), b.header.height)
+            }
+          )
+          .sequence
+
+      def execution(
+          validating: Ref[Task, Boolean],
+          executing: Ref[Task, Boolean],
+          achievedPar: Ref[Task, Boolean]
+      ) =
+        batch
+          .map(b =>
+            Task.defer {
+              for {
+                _ <- executing.set(true)
+                v <- validating.get
+                _ <- achievedPar.update(_ || v)
+                r <- checkpointingService.executeBlock(b, commitQC, commitPath)
+                _ <- executing.set(false)
+              } yield r
+            }
+          )
+          .sequence
+
+      val ledgerStorageCheck = store.runReadOnly {
+        ledgerStorage.get(finalLedger.hash)
+      }
+
+      for {
+        _ <- persistBlocks
+
+        // used to make sure that parallelism was achieved
+        validating  <- Ref[Task].of(false)
+        executing   <- Ref[Task].of(false)
+        achievedPar <- Ref[Task].of(false)
+
+        (validationRes, executionRes) <- Task.parZip2(
+          validation(validating, executing, achievedPar),
+          execution(validating, executing, achievedPar)
+        )
+
+        par             <- achievedPar.get
+        persistedLedger <- ledgerStorageCheck
+        ledgerTree      <- ledgerTreeRef.get
+      } yield {
+        val validationsAfterExec = validationRes.collect {
+          case (r, h) if h > batch.last.header.height => r
+        }
+
+        val ledgerTreeReset = batch.reverse match {
+          case committed :: rest =>
+            ledgerTree
+              .get(committed.hash)
+              .contains((finalLedger, committed.header)) &&
+              rest.forall(b => !ledgerTree.contains(b.hash))
+
+          case _ => false
+        }
+
+        //val validationsSaved =
+        //  validationBatch.diff(batch).forall(b => ledgerTree.contains(b.hash))
+
+        all(
+          "parallelism achieved" |: par,
+          "validation successful" |: validationsAfterExec.forall(identity),
+          "execution successful" |: executionRes.forall(identity),
+          "ledger persisted" |: persistedLedger.contains(finalLedger),
+          "ledgerTree reset" |: ledgerTreeReset
+          //TODO: possible optimisation
+          //  This test shows that upon committing a block the projected state of its descendants may be lost.
+          //  It's not techically wrong, as blocks will still validated correctly, but that state could potentially
+          //  be reused
+          //"ledgerTree contains validations" |: validationsSaved
```

Comment:
Good to have a comment and test case to show this :heart:

Reply:
I actually forgot to finish this comment 🙃 
I meant to note that a possible improvement is, upon `saveState`, not to remove any descendants (if found) of the committed block. Do you think we should do that?

Reply:
I would have personally tried to go that way, hence the generalisation of the tree storage in https://github.com/input-output-hk/metronome/pull/56, but I also like the simple assurance against memory leaks that a full reset provides. 

In practice it's hopefully not going to be the case that we have a long chain of uncommitted blocks building up that have to be re-entered into the ledger state all the time, but this optimisation seemed like something the flagship use case would do.

The important thing is that now we have these tests, so you refactoring can be done with confidence.

Reply:
Done, also solves the problem of validating against the parent ledger